[indexer-alt] add coin jsonrpc impl

emmazzz opened the pull request:
## Description 

This PR added indexer alt jsonrpc reader implementation for `getCoins` and `getAllCoins`. I ported over `raw_query` from graphql crate for usage here.

## Test plan 

Added e2e tests.

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] gRPC:
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

lxfind commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,241 @@
+            // let cursor_coin_query = cb::table.select(StoredCoinBalanceBucket::as_select()).filter(cb::object_id.eq(cursor.to_vec()));
+            let mut cursor_coins: Vec<StoredCoinBalanceBucket> =
+                conn.results_with_raw_query(cursor_coin_query).await?;
+            if cursor_coins.len() != 1 {
```
Not sure what you are trying to achieve here, but there can and will be multiple entries with the same object_id in this table, since it is an append table, we add entries for new mutations.

emmazzz replied:
I'm trying to get the balance bucket of the cursor coin here so that we can use the balance in the filter, because results are sorted by balance. And you are right I should account for the append-only-ness here. I've updated the query in the latest commit to select the row with the latest cp sequence number.

lxfind commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,241 @@
+
+        // First load the coins' object ids.
+        // TODO: optimize this query according to the indexes in coin_balance_buckets table.
+        let mut query = query!("
```
I think you would have to rewrite the entire function with a full raw string as the query.

wlmyng commented on crates/sui-indexer-alt-jsonrpc/src/reader.rs:
```diff
@@ -0,0 +1,231 @@
+            ));
+        }
+
+        let objects: Vec<StoredObject> = self.results_with_raw_query(query).await?;
+        Ok(objects)
```
saw you do it elsewhere: could we also get away with `self.results(sql_query(x))` where `(x, _) = query.finish()`? maybe we dont need both fns

wlmyng commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,284 @@
+            limit = limit + 1 // Query one more result to determine if there is a next page.
+        );
+
+        let buckets: Vec<IdCpBalance> = conn.results(sql_query(query)).await?;
```
it looks like we dont have any bindings so we're doing `conn.results(sql_query(query))` here - do we have any situations where we do make use of bindings? otherwise, maybe we can just write raw string queries instead of using `RawQuery`?

wlmyng commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,284 @@
+            .filter_map(|bucket| ObjectID::from_bytes(&bucket.object_id).ok())
+            .collect();
+
+        // Then we load the actual objects from kv, to render the coins.
+        let objects = conn.load_latest_objects(coin_ids.clone()).await?;
+
```
Not sure if this is too hacky, but with a dataloader, maybe we could do:
```
let coin_handles: Vec<_> = buckets
   .into_iter()
   .filter_map(|bucket| ObjectID::from_bytes(&bucket.object_id).ok())
   .map(|id| dataloader.load_one(id))
   .collect();

let sorted_objects = futures::future::join_all(coin_handles).await?;
```

lxfind commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,284 @@
+            );
+        }
+
+        if let Some(cursor) = cursor {
```
As we discussed, we will need at least the cp_sequence_number in the cursor.
Ideally also include balance bucket in the cursor

emmazzz commented:
Revamped and rebased this PR based on the latest changes in configs, pagination, errors and data loaders. 
Another noticeable change is with the new cursor format consisting of (object_id, cp_sequence_number, balance_bucket) @gegaowp .

emmazzz commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,336 @@
+                    ObjectID::from_bytes(&bucket.object_id).context("Failed to parse object id")?;
+                let balance_bucket = bucket
+                    .coin_balance_bucket
+                    .ok_or_else(|| anyhow::anyhow!("Coin balance bucket is null"))?;
```
I don't think I'm fully grasping how error handling should properly work here 😅
`.context` turns the error into an `anyhow::Error` so here I use `anyhow::anyhow`.
Would like @amnn 's advice on if this is the right way.

amnn commented on crates/sui-indexer-alt-e2e-tests/tests/jsonrpc/coin/deleted.move:
```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: Apache-2.0
+
+// This test is to verify that deleted coins are not included in the result of suix_getCoins.
+// We create two coins, of balances 12 and 34, call the rpc method to see bot of them in the results.
```
```suggestion
// We create two coins, of balances 12 and 34, call the rpc method to see both of them in the results.
```

amnn commented on crates/sui-indexer-alt-e2e-tests/tests/jsonrpc/coin/get_all_coins.exp:
```diff
@@ -0,0 +1,119 @@
+  "result": {
+    "data": [
+      {
+        "coinType": "0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI",
```
This output makes me think that we are not using the coin balance buckets index as efficiently as we could, because we are merging together different coin types, but IIRC coin type is the second field in the index after owner (so we should be returning all the results for one coin type in one go).

This also implies that the cursor should include the coin type. The cursor is also already quite large, so this may indicate that we need to introduce a BCS-encoded cursor (similar to what we did for GraphQL), but we can do that as a follow-up.

amnn commented on crates/sui-indexer-alt-e2e-tests/tests/jsonrpc/coin/inconsistency.move:
Note that this test is not really about inconsistency, it's about the fact that we only update `coin_balance_buckets` if a coin's bucket changes, or its owner changes, but we still want to get its up-to-date balance, which may be due to a transaction that appears strictly after the checkpoint that the bucket update came from. We would expect the same behaviour from GraphQL, the difference would be that in GraphQL there would be yet another field in the cursor, representing the `checkpoint_viewed_at`, and we would bound how far ahead in time we look for the latest balance by this value.

emmazzz replied:
I guess what I'm trying to show is that when you paginate in jsonrpc, the results on different pages might be from different checkpoints, while with GraphQL the results from all pages will be consistent due to `checkpoint_viewed_at`. I will update the test to reflect this better.

amnn commented on crates/sui-indexer-alt-e2e-tests/tests/jsonrpc/coin/pagination.move:
```diff
@@ -0,0 +1,89 @@
+  "params": ["@{A}", null, null, 3]
+}
+
+//# run-jsonrpc --cursors {"object_id":@{obj_2_0},"cp_sequence_number":1,"coin_balance_bucket":4}
```
In other places, when we interpolate variables, we don't include quotes -- it would be good to stick to that convention.

```suggestion
//# run-jsonrpc --cursors {"object_id":"@{obj_2_0}","cp_sequence_number":1,"coin_balance_bucket":4}
```

amnn reviewed (commented):
This is looking in pretty good shape, thanks @emmazzz. Here's a summary of the comments I left, from most to least important:

- Query performance -- I think the implementation of `getAllCoins` may have an unnecessary sort in it that we can avoid, and I'm a little bit worried about the DB's ability to efficiently plan the comparison on bucket, checkpoint, ID (so worth double checking that).
  - There's a tangent here about the size of the cursor and potentially using BCS encoded cursors which we can discuss.
- Transactional test runner change -- this is a bit of a larger change and needs some more design -- I've proposed something in the inline comments. The issue with the current approach is that the feature being introduced (interpolating object IDs into cursors) bridges two places where we use similar syntax for different things, which is likely to cause confusion.
- Errors -- I think are mostly there, but I made some suggestions to bring them fully in line (the biggest thing to address here is the conversion function not using `unwrap`s, the second biggest thing is the behaviour when we fail to fetch contents which we need to align on).

emmazzz commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,336 @@
+    Deserialization(#[from] bcs::Error),
+}
+
+#[derive(Queryable, Selectable, Debug, Serialize, Deserialize)]
```
You are right `allow_unused` is no longer needed now that we have `cp_sequence_number` in the cursor. Looks like I do need a dedicated type to make diesel happy though, moving it to the function.

emmazzz commented on crates/sui-transactional-test-runner/src/test_adapter.rs:
```diff
@@ -1352,7 +1350,28 @@ impl<'a> SuiTestAdapter {
                 variables.insert(format!("cursor_{idx}"), base64d);
             } else {
-                variables.insert(format!("cursor_{idx}"), Base64::encode(s));
+                // Handle cursor strings containing @{obj_x_y} pattern
```
> Support general interpolation of variables into the cursor body -- there's already a function that replaces variables with their contents, so you should be able to re-use that.

Right there's an `interpolate_query` function that does variable replacement but it seems to treat cursor contents as part of the variable map because we indeed use variables like `@{cursor_0}` in the actual contents and require knowing the value of `@{cursor_0}` beforehand. Anyway I think this function warrants some refactoring and cleanup. Do you think I could do this in a followup PR?

> Pick new syntax for the old @{obj_x_y} and @{obj_x_y,n} cursors. Maybe just prefix them with bcs:? (e,g. bcs:@{obj_x_y}, bcs:@{obj_x_y,n})

Good idea to differentiate the cursor formats. This can be done in the followup PR too.

amnn replied:
Recording our conversation from the pair programming session: It's okay (beneficial, even) to only interpolate other variables when dealing with interpolation for cursors (i.e. pass in an empty list of cursors), and I think it's a good idea to factor this change out into its own commit, at least, but that commit should come first (because the test adapter is being actively touched by multiple people at the moment, so I would rather not increase the confusion factor there).

emmazzz commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,336 @@
+            query = query
+                .filter(candidates!(cp_sequence_number).le(cp_sequence_number as i64))
+                .filter(
+                    // Since the result is ordered by coin_balance_bucket followed by object_id,
```
I did `EXPLAIN ANALYZE` the query and the query plan looks like this:
```
Incremental Sort  (cost=52.82..977.98 rows=87 width=43)
   Sort Key: candidates.coin_balance_bucket DESC, candidates.object_id
   Presorted Key: candidates.coin_balance_bucket
   ->  Nested Loop Anti Join  (cost=1.52..974.59 rows=87 width=43)
         ->  Index Only Scan using coin_balances_buckets_owner_type on coin_balance_buckets candidates  (cost=0.82..383.31 rows=131 width=43)
               Index Cond: ((owner_kind = 0) AND (owner_id = '\x885672e6744326979bd77e03d769094746103a8a0dbd0710d8e542840ee048ff'::bytea) AND (coin_type = '\x070000000000000000000000000000000000000000000000000000000000000002037375690353554900'::bytea) AND (coin_balance_bucket IS NOT NULL) AND (cp_sequence_number <= 80872138))
               Filter: ((coin_balance_bucket < 10) OR ((coin_balance_bucket = 10) AND (object_id > '\x7832636331656238643234666461383638646362626636393537356234653764373932656639646236373566323962306561393631656639373832323536363233'::bytea)))
         ->  Index Only Scan using coin_balance_buckets_pkey on coin_balance_buckets newer  (cost=0.70..4.75 rows=1 width=41)
               Index Cond: ((object_id = candidates.object_id) AND (cp_sequence_number > candidates.cp_sequence_number))
(9 rows)
```
which looks fine to me?
One issue I did notice while running the original query though was that `owner_kind` needed to be in the query because it is the first column in the index for some reason. Without `owner_kind` in the query the plan is slow and uses sequential scan. This owner kind corresponds to `Fastpath` or `Consensus`. I've changed the query to filter by `owner_kind = Fastpath` but I don't think that's right since we probably want to show consensus objects too. But then how do we make sure the index is used efficiently?

amnn replied:
We talked about how we would probably need to invert the direction of object IDs in the index to make this as efficient as possible.

amnn commented on crates/sui-indexer-alt-jsonrpc/src/lib.rs:
```diff
@@ -5,10 +5,12 @@ use std::net::SocketAddr;
 use api::rpc_module::RpcModule;
 use api::transactions::{QueryTransactions, Transactions, TransactionsConfig};
 use config::RpcConfig;
+
```
nit: random whitespace (this will confuse/mess up rust-analyzer).

amnn reviewed (commented):
Recording our conversation from today!

emmazzz commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,349 @@
+        })) = page.cursor
+        {
+            query = query.filter(sql::<Bool>(&format!(
+                r#"ROW(candidates.coin_balance_bucket, candidates.cp_sequence_number, candidates.object_id)
```
Changed to using `ROW` here and ordering by `coin_balance_bucket, cp_sequence_number, object_id `all DESC a few lines later as we've discussed @amnn

amnn replied:
Rather than using `&format!(...)`, use diesel's `bind` construct, so that we don't have to worry about re-implementing escaping/encoding. See below from #21119 for an example:

https://github.com/MystenLabs/sui/blob/bc3b4c602d3dbd88b10776f3a23b09477a3a1b7c/crates/sui-indexer-alt-jsonrpc/src/api/objects/filter.rs#L155-L163

(I only learned about this recently, while working on this PR). It would look something like this for `coin_balance_buckets`:

```rust
if let Some(Cursor(c)) = page.cursor {
    query = query.filter(
        sql::<Bool>(r#"("candidates"."coin_balance_bucket", "candidates"."cp_sequence_number", "candidates"."object_id") < ("#)
            .bind::<BigInt, _>(c.coin_balance_bucket as i64)
            .sql(", ")
            .bind::<BigInt, _>(c.cp_sequence_number as i64)
            .sql(", ")
            .bind::<Bytea, _>(c.object_id.clone())
            .sql(")"),
    );
}
```

(the `ROW` keyword is optional).

emmazzz commented:
Updated the PR to address the second round of comments, include the index change on `coin_balance_bucket` and also rebased on top of the transactional test runner cursor pr #21150

amnn commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,349 @@
+        coin_type: Option<String>,
+        cursor: Option<String>,
+        limit: Option<usize>,
+    ) -> RpcResult<JsonRpcPage<Coin, String>> {
```
nit: for consistency with other implementations of paginated methods (and also generally with the naming scheme for response types generally).

```suggestion
    ) -> RpcResult<PageResponse<Coin, String>> {
```

amnn commented on crates/sui-indexer-alt-schema/migrations/2025-02-05-232254_coin_balance_buckets_object_id_desc/down.sql:
Let's split the index changes out into their own PR -- unlike the rest of this PR, we can't land them until we have run the migration on our production instances, which may need to wait until a gap in the various benchmarks.

amnn commented on crates/sui-indexer-alt-schema/migrations/2025-02-05-232254_coin_balance_buckets_object_id_desc/down.sql:
```diff
@@ -0,0 +1,4 @@
+DROP INDEX IF EXISTS coin_balances_buckets_owner_type;
+
+CREATE INDEX IF NOT EXISTS coin_balances_buckets_owner_type
```
Name the new and old indices differently to each other, otherwise these migrations will always delete whatever index exists and recreate it (even if it had the correct structure). We want to be able to run the steps in this migration manually on the production instance (once we have the all clear), and then have this script detect that those steps have run.

amnn commented on crates/sui-indexer-alt-schema/src/objects.rs:
nit: spurious?

amnn commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,349 @@
+    }
+}
+
+impl Coins {
```
nit: it would be nice if we could structure this endpoint like the rest of them (with a `response` module to hold these functions). The rest of the methods are structured that way to promote sharing (it's not possible to share response functions when they are tied to the type implementing the method), and it would be good to maintain consistency so that people can follow that pattern when they are navigating the codebase.

emmazzz replied:
I thought about it but decided not to do it that way because:
- I don't think the code here is long or complex enough to need to be split up into different modules. In my opinion being in different files and modules can hurt readability (personally I don't like jumping between files), unless the code is long (`transactions`'s case) or shareable (`objects'`'s case), in which case the benefits outweigh the drawbacks, which bring me to the second point.
- I don't see the response function in this case being shared. The only methods that return coin related responses are the methods in this file, `getCoins`, `getAllCoins` and `getAllBalances`. And we certainly don't plan to expand jsonrpc interface so most likely no future methods will be using it.
Or do you think this needs to be done for consistency reasons, i.e. any method that has filters and pagination should follow that structure?

amnn commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,349 @@
+            )));
+        }
+
+        // Finally we order by coin_balance_bucket and break ties by object_id.
```
nit: comment needs updating

amnn commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,349 @@
+            .select(candidates!(
+                object_id,
+                cp_sequence_number,
+                coin_balance_bucket
```
Did you try adding `assume_not_null` here? (based on the discussion below around `IdCpBalance` needing to use an optional field for this?)

amnn commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,349 @@
+                    ObjectID::from_bytes(&bucket.object_id).context("Failed to parse object id")?;
+                let balance_bucket = bucket
+                    .coin_balance_bucket
+                    .ok_or_else(|| anyhow::anyhow!("Coin balance bucket is null"))?;
```
You probably won't need it in this case (if you set `assume_not_null` above) but I recently learned that `anyhow`'s `context` works on optional values as well, so you can do:

```suggestion
                    .context("Coin balance bucket is null")?;
```

amnn commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,349 @@
+            .last()
+            .map(|bucket| {
+                let object_id =
+                    ObjectID::from_bytes(&bucket.object_id).context("Failed to parse object id")?;
```
FWIW in GraphQL, we tend to use the bytes from the DB in the cursor, to avoid having to deserialize and then reserialize values here. I did the same thing for the owned object cursor, and it simplifies this code a fair bit:

https://github.com/MystenLabs/sui/blob/bc3b4c602d3dbd88b10776f3a23b09477a3a1b7c/crates/sui-indexer-alt-jsonrpc/src/api/objects/filter.rs#L197-L207

It also reduces the need for defining a dedicated type to represent the output of this query, because it maps exactly onto the cursor type.

amnn commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,349 @@
+
+    async fn coin_response(&self, id: ObjectID) -> Result<Coin, RpcError<Error>> {
+        let Self(ctx, _) = self;
+        let stored_object = load_latest(ctx.loader(), id)
+            .await?
+            .ok_or_else(|| anyhow::anyhow!("Failed to load latest object {:?}", id))?;
+
+        let object: Object =
+            bcs::from_bytes(&stored_object.serialized_object.ok_or_else(|| {
+                anyhow::anyhow!("Failed to deserialize object {:?}", stored_object.object_id)
+            })?)
+            .context("Failed to deserialize object")?;
```
#21161 includes a refactoring such that `load_latest` always returns an `Object`.

amnn commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,349 @@
+        let Self(ctx, _) = self;
+        let stored_object = load_latest(ctx.loader(), id)
+            .await?
+            .ok_or_else(|| anyhow::anyhow!("Failed to load latest object {:?}", id))?;
```
Why use the debug output format for `ObjectID` here? the `Display` format should do the right thing.

amnn commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,349 @@
+        let digest = object.digest();
+        let version = object.version();
+        let previous_transaction = object.as_inner().previous_transaction;
+        let coin = object
+            .as_coin_maybe()
+            .ok_or(anyhow::anyhow!("Object is expected to be a coin"))?;
+        let coin_type = object
+            .coin_type_maybe()
+            .ok_or(anyhow::anyhow!("Object is expected to have a coin type"))?
+            .to_canonical_string(/* with_prefix */ true);
```
```suggestion
        let coin = object
            .as_coin_maybe()
            .context("Object is expected to be a coin")?;
        let coin_type = object
            .coin_type_maybe()
            .context("Object is expected to have a coin type")?
            .to_canonical_string(/* with_prefix */ true);
```

amnn commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,349 @@
+            .await
+            .into_iter()
+            .zip(coin_id_page.data)
+            .map(|(r, c)| {
+                let id = c;
+                r.with_internal_context(|| format!("Failed to get object {id}"))
+            })
```
```suggestion
            .map(|(r, id)| {
                r.with_internal_context(|| format!("Failed to get object {id}"))
            })
```

amnn commented on crates/sui-indexer-alt-e2e-tests/tests/jsonrpc/coin/get_all_coins.snap:
```diff
@@ -0,0 +1,122 @@
+  "result": {
+    "data": [
+      {
+        "coinType": "0x0000000000000000000000000000000000000000000000000000000000000002::sui::SUI",
```
The ordering of this response still does not seem correct -- we need to order by coin type first, then bucket, etc. Did you take a look at the query plan for `getAllCoins`?

amnn commented on crates/sui-indexer-alt-jsonrpc/Cargo.toml:
We would no longer need this dependency if we use `SqlLiteral::bind`.

amnn reviewed (commented):
Almost there -- main comments:

- I suspect `getAllCoins` is still not going to perform well because of its ordering. We need to order by coin type first, then balance bucket, etc
- The index changes need to be split into their own PR to unblock landing this one (also some suggestions for making the migration idempotent and easy to run proactively).
- Various patterns/suggestions based on #21119. Main ones are to use `SqlLiteral::bind`, represent the ObjectID using its DB form in the cursor, and to use `context(...)` instead of `.ok_or_else(|| anyhow!(...))`.

emmazzz commented:
Hey @amnn thanks for the review! I have addressed most of the comments in the latest commit. Specifically,
- `getAllCoins` has been removed. I agree with your assessment and I think we can tackle this one after Stashed milestone.
- The index change has been split into its own PR #21187
- I got `assume_not_null` working but I still need to have a separate struct for cursor for the time being because changing the object id type into `Vec<u8>` of the json cursor type will make it hard to test, given how we handle json cursor and object id replacement in test adapter. We are going to make it a bcs cursor anyway later when bcs cursor lands. I have added TODOs for it in the code.

I understand there are changes in other PRs that can further improve this one (bcs cursor and your improvements to `load_latest`) but I think these improvements are minor and don't need to block the landing of this one. What do you think?

amnn commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,322 @@
+        // Construct the basic query first to filter by owner, not deleted and newest rows.
+        let mut query = candidates
+            .select((
+                candidates.field(cb::object_id),
+                candidates.field(cb::cp_sequence_number),
+                candidates.field(cb::coin_balance_bucket).assume_not_null(),
```
Would the following work?

```suggestion
                candidates!(object_id),
                candidates!(cp_sequence_number),
                candidates!(coin_balance_bucket).assume_not_null(),
```

amnn commented on crates/sui-indexer-alt-jsonrpc/src/api/coin.rs:
```diff
@@ -0,0 +1,322 @@
+        })) = page.cursor
+        {
+            query = query.filter(
+                sql::<Bool>(r#"(candidates.coin_balance_bucket, candidates.cp_sequence_number, candidates.object_id) < ("#)
```
You can drop the raw string if the string contents does not include quotes:
```suggestion
                sql::<Bool>("(candidates.coin_balance_bucket, candidates.cp_sequence_number, candidates.object_id) < (")
```

amnn commented on crates/sui-indexer-alt-e2e-tests/tests/jsonrpc/coin/pagination.move:
```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// This test is to verify that deleted coins are not included in the result of suix_getCoins.
+// We create two coins, of balances 12 and 34, call the rpc method to see bot of them in the results.
+// Then we merge the coins and call the rpc method again to see that only the merged coin with
+// balance 46 is in the results.
+
```
nit: comment copypasta

amnn reviewed (approved):
Some small nits, but this is good to go! Agreed we can incorporate BCS cursors in a follow-up.
